tests/unit/data/data_pipeline/test_parquet_dataloader.py

```diff
@@ -0,0 +1,299 @@
+
+import numpy as np
+import numpy.typing as npt
+import pandas as pd
+import pyarrow as pa
+import pyarrow.parquet as pq
+
+from fairseq2.utils.parquet_dataloader import (
+    ParquetBasicDataLoader,
+    ParquetBasicDataloaderConfig,
+    ParquetBatchFormat,
+)
+
+
+def gen_random_string(length: int) -> str:
+    return "".join(
+        random.choice(string.ascii_letters + string.digits) for n in range(length)
+    )
+
+
+def generate_random_pandas_df(size: int, seed: int = 123) -> pd.DataFrame:
+    np_rs = np.random.RandomState(seed)
+    df: tp.Dict[str, tp.Union[npt.NDArray[tp.Any], tp.List[tp.Any]]] = {}
+    df["int_col"] = np_rs.randint(0, 200, size)
+    df["float_col"] = np_rs.randn(size)
+
+    df["string_col1"] = [gen_random_string(10) for _ in range(size)]
+    df["string_col2"] = [gen_random_string(2) for _ in range(size)]
+
+    df["list_int_col"] = [
+        np_rs.randint(-10, 10, np_rs.randint(0, 100)) for _ in range(size)
+    ]
+    df["list_float_col"] = [
+        np_rs.rand(np_rs.randint(0, 10)).astype(np.float32) for _ in range(size)
+    ]
+    df["list_float_fixed_size_col"] = [
+        np_rs.rand(7).astype(np.float32) for _ in range(size)
+    ]
+    return pd.DataFrame(df)
+
+
+def generated_partitioned_parquet_file(
+    path: str, size: int, n_partitions: int = 20, seed: int = 123
+) -> None:
+    df = generate_random_pandas_df(size, seed)
+
+    if n_partitions > 0:
+        df["part_key"] = np.arange(size) % n_partitions
+
+    table = pa.Table.from_pandas(df)
+
+    pq.write_to_dataset(
+        table,
+        path,
+        partition_cols=["part_key"] if n_partitions > 0 else None,
+        existing_data_behavior="delete_matching",
+        **{"row_group_size": 110},
+    )
+
+
+class TestParquetDataloader(unittest.TestCase):
+    _tmpdir: str
+    _tmp_parquet_ds_path: str
+    _tmp_parquet_single_path: str
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        cls._tmpdir = tempfile.mkdtemp()
+        cls._tmp_parquet_ds_path = os.path.join(cls._tmpdir, "test")
+        generated_partitioned_parquet_file(cls._tmp_parquet_ds_path, size=2 * 10**3)
+
+        cls._tmp_parquet_single_path = os.path.join(cls._tmpdir, "single_test.parquet")
+        generated_partitioned_parquet_file(
+            cls._tmp_parquet_single_path, size=10**3, n_partitions=0
+        )
+
+    @classmethod
+    def tearDownClass(cls) -> None:
+        # Cleanup temp working dir.
+        if cls._tmpdir is not None:
+            shutil.rmtree(cls._tmpdir)  # type: ignore
+
+    def test_simple_dataload(self) -> None:
+        config = ParquetBasicDataloaderConfig(
+            parquet_path=self._tmp_parquet_ds_path,
+            batch_size=11,
+            nb_producers=2,
+            seed=333,
+        )
+        pbdl = ParquetBasicDataLoader(config)
+        res: tp.List[pd.DataFrame] = list(iter(pbdl))
+
+        for x in res:
+            self.assertIsInstance(x, pa.Table)
+
+        self.assertEqual(
```

Comment:
in pytest this just become `assert ... == ...`

Reply:
i'm staying in UnitTest conventions since I started to use them

Reply:
since you're trying to merge a PR to fairseq2, it would be better to follow the style and conventions used there.